integration/data/test_restore_to.py

```diff
@@ -0,0 +1,252 @@
+import os
+import pyqcow
+
+import cmd
+import common
+from common import replica1, replica2, controller, backing_replica1, backing_replica2, backup_targets  # NOQA
+from common import read_dev, read_file, read_from_backing_file
+from utils import file, create_backup, rm_backups, rm_snaps
+
+VOLUME_NAME = 'test-volume_1.0'
+VOLUME_SIZE = str(4 * 1024 * 1024)  # 4M
+BLOCK_SIZE = str(2 * 1024 * 1024)  # 2M
+
+BACKING_FILE_QCOW2 = 'backing_file.qcow2'
+
+OUTPUT_FILE_RAW = 'test-restore_to-output.raw'
+OUTPUT_FILE_QCOW2 = 'test-restore_to-output.qcow2'
+
+IMAGE_FORMAT_RAW = 'raw'
+IMAGE_FORMAT_QCOW2 = 'qcow2'
+
+
+def check_backing(dev, offset=0, length=4*1024):
+    backing_data = read_from_backing_file(offset, length)
+    empty_volume_data = read_dev(dev, offset, length)
+    assert backing_data == empty_volume_data
+
+
+def read_qcow2_file_without_backing_file(file, offset, length):
+    assert os.path.exists(file)
+    qcow_file = pyqcow.file()
+    qcow_file.open(file)
+    qcow_file.seek(offset)
+    return qcow_file.read(length)
+
+
+def snapshots_with_backing_file_test(backing_dev, backup_target):
+    length0 = 4 * 1024
+    length1 = 256
+    length2 = 128
+    offset0 = 0
+    offset1 = length1 + offset0
+    offset2 = length2 + offset0
+
+    output_raw_path = file(OUTPUT_FILE_RAW)
+    output_qcow2_path = file(OUTPUT_FILE_QCOW2)
+
+    # create 1 empty snapshot.
+    # data in output image == data in backing file
+    check_backing(backing_dev)
+    snap0 = cmd.snapshot_create()
+    backup = create_backup(backup_target, snap0)
+
+    init_data = read_dev(backing_dev, offset0, length0)
+    backing_data = read_from_backing_file(offset0, length0)
+    assert init_data != ""
+    assert init_data == backing_data
+
+    cmd.restore_to(backup, file(BACKING_FILE_QCOW2),
+                   output_raw_path, IMAGE_FORMAT_RAW)
+    output0_raw = read_file(output_raw_path, offset0, length0)
+    assert output0_raw == backing_data
+    os.remove(output_raw_path)
+    assert not os.path.exists(output_raw_path)
+
+    cmd.restore_to(backup, file(BACKING_FILE_QCOW2),
+                   output_qcow2_path, IMAGE_FORMAT_QCOW2)
+    output0_qcow2 = read_qcow2_file_without_backing_file(
+        output_qcow2_path, offset0, length0)
+    assert output0_qcow2 == backing_data
+    os.remove(output_qcow2_path)
+    assert not os.path.exists(output_qcow2_path)
+
+    rm_backups(backup)
+
+    # create 1 snapshot with 256B data.
+    # output = snap1(offset0, length1) + backing(offset1, ...)
+    snap1_data = common.random_string(length1)
+    common.verify_data(backing_dev, offset0, snap1_data)
+    snap1 = cmd.snapshot_create()
+    backup = create_backup(backup_target, snap1)
+
+    backing_data = read_from_backing_file(
+        offset1, length0 - offset1)
+
+    cmd.restore_to(backup, file(BACKING_FILE_QCOW2),
+                   output_raw_path, IMAGE_FORMAT_RAW)
+    output1_raw_snap1 = read_file(
```

Comment:
You should also to compare the result with what you read from the device directly.

Same below.